proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,408 @@
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+= 67% - 5% - 20% = 42%, so slots with above 42% votes accumulated from
+`RestartLastVotedForkSlots` would be repaired.
+
+From above examples, we can see the "must-have" threshold changes dynamically 
+depending on how many validators are in restart. The main benefit is that a
+block will only move from "must-have" to "ignored" as more validators 
+join the restart, not vice versa. So the list of blocks a validator needs to
+repair will never grow bigger when more validators join the restart.
+
+Once the validator gets `RestartLastVotedForkSlots`, it can calculate which
+blocks must be repaired. When all those "must-have" blocks are repaired and
+replayed, it can proceed to step 3.
+
+3. **gossip current heaviest fork**
+
+The main goal of this step is to "vote" the heaviest fork to restart from.
+
+We use a new gossip message `RestartHeaviestFork`, its fields are:
+
+* `slot`: `u64` slot of the picked block.
+* `hash`: `Hash` bank hash of the picked block.
+* `stake_committed_percent`: `u16` total percentage of stakes of the validators
+it received `RestartHeaviestFork` messages from.
+
+After receiving `RestartLastVotedForkSlots` from the validators holding stake
+more than `RESTART_STAKE_THRESHOLD` and repairing slots in "must-have"
+category, replay all blocks and pick the heaviest fork by traversing from
+local root like this:
+
+1. If a block has more than 67% stake in `RestartLastVotedForkSlots`
+   messages, traverse down this block.
+
+2. Define the "must have" threshold to be 62%. If traversing to a block with
+   more than one child, we check for each child
+   `vote_on_child + stake_on_validators_not_in_restart >= 62%`. If so, traverse
+   to the child.
+
+   For example, if 80% validators are in restart, child has 42% votes, then
+   42 + (100-80) = 62%, traverse to this child. 62% is chosen instead of 67%
+   because 5% could make the wrong votes.
+
+   Otherwise stop traversing the tree and use last visited block.
+
+To see why the above algorithm is safe, assuming one block X is optimistically
+confirmed before the restart, we prove that it will always be either the block
+picked or the ancestor of the block picked.
+
+Assume X is not picked, then:
+
+1. If its parent Y is on the Heaviest fork, then either a sibling of X is
+   chosen or no child of Y is chosen. In either case
+   vote_on_X + stake_on_validators_not_in_restart < 62%, otherwise X would be
+   picked. This contradicts with the fact X got 67% votes before the restart,
+   because vote_on_X should have been greater than
+   67% - 5% (non-conforming) - stake_on_validators_not_in_restart.
+
+2. If its parent Y is also not on the Heaviest fork, Y should have got > 67%
+   of the votes before the restart as well, then we can apply the same
+   reasoning to Y's parent, until we find an ancestor which is on the Heaviest
+   fork, then the contradiction in previous paragraph applies.
```

Comment:
Don't think we need to mention an inductive argument here, since we specified `X` is any arbitrary optimistic confirmed slot

```suggestion
2. If its parent Y is also not on the Heaviest fork, Y should have got > 67%
   of the votes before the restart as well, then we can apply the same
   reasoning to Y's parent, until we find an ancestor which is on the Heaviest
   fork, then the contradiction in previous paragraph applies.
```

Reply:
Changed.